### DISCUSSION TITLE
TEDAPI - Local API for Powerwall

### DISCUSSION THREAD

**Discussion by jasonacox** [jasonacox/pypowerwall#104] [General]
How to use the latest pypowerwall to connect to the Powerwall gateway TEDAPI interface:

First, you need to create a network route to get to that endpoint via your local network (or direct connect to the Gateway's WiFi access point):

```bash
# Change 192.168.0.100 to the lP address of the Powerwall gateway on your LAN

# Linux Ubuntu and RPi - Can add to /etc/rc.local for persistence 
sudo ip route add 192.168.91.1 via 192.168.0.100

# MacOS 
sudo route add -host 192.168.91.1 192.168.0.100 # Temporary 
networksetup -setadditionalroutes Wi-Fi 192.168.91.1 255.255.255.255 192.168.0.100 # Persistent

# Windows - Using persistence flag - Administrator Shell
route -p add 192.168.91.1 mask 255.255.255.255 192.168.0.100
```

Next, find the Gateway WiFi password that is typically located on the QR code on the gateway itself. Use that to connect with something like this:

```python
Import pypowerwall

PW_HOST=“192.168.91.1"
PW_GW_PWD=“GATEWAY_PWD"

pw = pypowerwall.Powerwall(PW_HOST, gw_pwd=PW_GW_PWD)

print(f"mode: {pw.mode} - tedapi: {pw.tedapi_mode}")
pw.tedapi.get_din()
pw.power()
```

Research on TEDAPI: https://github.com/jasonacox/Powerwall-Dashboard/discussions/392

**Comment by JohnJ9ml** [comment]
Checking in some details here:

My installation consists of one Gateway 3 and three Powerwall 3 units. The Tesla One account is connected to my primary PW3. I looked and I don't see any Wifi information or the like on the Gateway 3. The PW3 is hardwired and on the network as 192.168.90.2. 

The Tesla One account shows the PW3 connected on cellular, Wifi, and Ethernet. 

I cite all of this because the instructions here say to connect to the gateway, but I don't see that information. My primary PW3 appears to be handling that role, and I'm not sure if that's new, evolutionary, etc.

When I try to use the tedapi (option 4), connecting to the primary PW3 with the Wifi password as the gateway password gives me back an Access Denied.

I'm wondering if things have changed here and the comments have not yet caught up, 

Running version 0.11.1.

**Reply by jasonacox** [reply]
HI @JohnJ9ml - I can't tell from your details, but are you connecting to 192.168.90.2 directly over your LAN? If so, that is the problem and you will get access denied because the TEDAPI requires you to be connected to the WiFi access point (not hardwired) at 192.168.91.1. 

Since your PW3 is on your LAN, you can create a static route (via 192.168.90.2) to get to that endpoint using a script like this:

https://github.com/jasonacox/Powerwall-Dashboard/blob/main/add_route.sh

Once you have that, your host should be able to ping 192.168.91.1 and you can run the code above to access your system.

Also, there is a built in test:

```bash
python3 -m pypowerwall.tedapi
```

**Reply by JohnJ9ml** [reply]
I apologize for not fully internalizing what was being said here. 

Somehow, my system being hardwired just didn't seem to convince me I really needed to find the IP address of the Wifi connection - it almost seems a bit unfathomable that the tedapi endpoint is evidently not exposed on a hardwired connection? 

I am now connected. One very small pickup from the readme for pypowerwall. For the Programming with pyPowerwall section, the tedapi example has a typo in the gateway password option:

pw = pypowerwall.Powerwall(host,password,email,timezone,gw_pw=gw_pw,auto_select=True)

Should be gw_pwd

It's right in the quick test code elsewhere, but wrong in the readme.

Thanks for the help and for getting me through one obtuse concept. It still seems a bit strange....

**Reply by jasonacox** [reply]
Thanks for the tip!  I updated the README: https://github.com/jasonacox/pypowerwall/commit/2c93ce9327d97020cb1ff78862ab9c38c8006e09

**Comment by cykeler** [comment]
Hello, not sure if this is the place to ask, but I keep having dropouts in data.  I am running pypowerwall dashboard via a Debian LXC container in proxmox. I set up the route using the add_route.sh script. I'm not sure why it keeps having connection issues.

if I show the route, its there. 
192.168.91.0/24 via 10.0.0.67 dev eth0

looking at the log it had connection issues
02/26/2025 08:55:48 AM [pypowerwall.tedapi] [ERROR] Error fetching status: HTTPSConnectionPool(host='192.168.91.1', port=443): Max retries exceeded with url: /tedapi/v1 (Caused by ConnectTimeoutError(<urllib3.connection.HTTPSConnection object at 0x7ca15220d780>, 'Connection to 192.168.91.1 timed out. (connect timeout=5)'))

after trying to ping the 192. address it returned nothing, I pinged the 10.0 address and it seemed to come back to life. Not sure if there is anything I need to do to keep it alive?

thanks! 

<img width="1452" alt="Screenshot 2025-02-26 at 8 51 39 AM" src="https://github.com/user-attachments/assets/1711d425-2f63-45ef-b2f8-a8a103105f72" />

**Reply by jasonacox** [reply]
I had dropout issues until I hardwired all the devices (Powerwall Gateway and Raspberry Pi in my case). Are you on WiFi?

If pinging always works to keep the ARP tables working, you could also set up a scheduled tasks (cron) that does something like a 5 ping:

```bash
ping -c 5 10.0.0.67
````

**Reply by cykeler** [reply]
Thanks for the response. The machine running proxmox is hardwired, but the pw3 is not. It's strange, I seem to have ping 8.8.8.8 before pinging the pw3 works again. I'll see if a cron to ping works. maybe I'll run it every few minutes.  While we are here, is there a way to edit the data source names ? i.e. change strings ABCDE to MPPT1 MPPT2 MPPT3, etc.
This is such an awesome setup, thanks so much!

**Reply by jasonacox** [reply]
Thanks @cykeler ! 🙏 

If you need to ping the internet from the proxmox that seems a bit more like an issue with your network router? 

The nice thing about Grafana is that every panel (graph) can be customized. Just edit the panels you want to change:

<img width="723" alt="image" src="https://github.com/user-attachments/assets/246000c0-a315-443f-878c-8aaa4748b4be" />

You can edit the name (alias) 

<img width="815" alt="image" src="https://github.com/user-attachments/assets/d2420318-80ba-4e6a-b489-5f91048ad156" />

**Reply by cykeler** [reply]
I'm not sure how I missed that edit in the dropdown. perfect!  I think you are right on it being a proxmox issue. Do you know if multiple devices can be accessing the pw3 ? I have teslemetry running via homeassistant, but that is using the fleet API.  FWIW I set up a cron to ping 5 every 5 minutes and it doesn't look like I had any dropouts through the night.
Is there anything I need to worry about if I am pinging the pw3 that much?

**Reply by cykeler** [reply]
sorry to bug. I did have one more question, I don't seem to be getting any data on the powerwall temps, any hints where to look to fix that?

**Reply by jasonacox** [reply]
> mutliple devices

I have several. :) all work fine.

> ping 5 every 5 minutes

Ping is a super tiny payload. No worries with this at all.

> don't seem to be getting any data on the powerwall temps

Unfortunately, the PW3 does not provide ambient temp data.

**Comment by Highland-Horatio** [comment]
Hope you can help.

Struggling to connect to my PW3 from either my PC or rpi using tedapi. My PW3 seems to have registered itself on my primary SSID as 192.168.1.115 but also appears as an additional SSID, TeslaPW_XXXXXX (last 6 chars redacted). If I connect direct to the Tesla SSID, the command "python -m pypowerwall tedapi" works, once I input the password and I get various data re my PW3, viz:

```
PS C:\Users\horat> python -m pypowerwall tedapi
pyPowerwall - Powerwall Gateway TEDAPI Reader [v0.13.2]
 - Connecting to https://192.168.91.1... SUCCESS

Enter Powerwall Gateway Password: XXXXXXXXXXX <redacted>

Connecting to Powerwall Gateway 192.168.91.1

 - Configuration:
   - Site Name: Todday Beag
   - Battery Commission Date: 2024-12-30T13:35:10+00:00
   - VIN: <redacted>
   - Number of Powerwalls: 1

 - Power Data:
   - Battery Charge: 91.64% (13150Wh of 14350.000000000002Wh)
   - Battery: -542W
   - Site: -3W
   - Load: 325W
   - Solar: 873W
   - Solar_Rgm: 0W
   - Generator: 0W
   - Conductor: 0W

 - Configuration and Status saved to config.json and status.json

```
However, the command does NOT work if I connect to my primary SSID - it cannot find the powerwall. I believed I therefore needed to create a route. Which I did. This time, the command seems to find powerwall but, after prompting me for the password, gives a "Failed to Connect" error message, Viz:

```
PS C:\Windows\system32> route add 192.168.91.1 mask 255.255.255.255 192.168.1.115
 OK!
PS C:\Windows\system32> ping 192.168.91.1

Pinging 192.168.91.1 with 32 bytes of data:
Reply from 192.168.91.1: bytes=32 time=53ms TTL=64
Reply from 192.168.91.1: bytes=32 time=2ms TTL=64
Reply from 192.168.91.1: bytes=32 time=13ms TTL=64

Ping statistics for 192.168.91.1:
    Packets: Sent = 3, Received = 3, Lost = 0 (0% loss),
Approximate round trip times in milli-seconds:
    Minimum = 2ms, Maximum = 53ms, Average = 22ms
Control-C
PS C:\Windows\system32> python -m pypowerwall tedapi
pyPowerwall - Powerwall Gateway TEDAPI Reader [v0.13.2]
 - Connecting to https://192.168.91.1... SUCCESS

Enter Powerwall Gateway Password: XXXXXXXXXXX <redacted>

Connecting to Powerwall Gateway 192.168.91.1
Access Denied: Check your Gateway Password
Failed to connect to Powerwall Gateway

ERROR: Unable to connect to Powerwall Gateway. Check your password and try again
PS C:\Windows\system32>

```
My supposition is that the password doesnt seem to be getting through the route? (I tried several times and checked I input the password correctly very carefully.) 

The above is a session on my PC, but the same seems to hold true from trying the same thing on my raspberrypi. 

PW3 running firmware 25.18.1
Python version: 3.13.5

Can you shed any light please?

Regards,
Harry.

**Reply by jasonacox** [reply]
Unfortunately, as of Powerwall Firmware version 25.10.0, network routing to the TEDAPI endpoint (192.168.91.1) is no longer supported by Tesla. They block this access. You must connect directly to the Powerwall's WiFi access point to access TEDAPI data. If you previously set up a static route for TEDAPI, you will need to remove it and connect directly to the WiFi access point as you discovered. I should add this note to the README.

FYI - some background https://github.com/jasonacox/Powerwall-Dashboard/discussions/645

**Reply by Highland-Horatio** [reply]
Hi Jason, Thanks for getting back to me and for the handy info - yep, a note on the README would be helpful for newbies following in my footsteps. Tesla are really are putting big fences in our way! The background to this is that I have been using Netzero to monitor and control my PW3 but Netzero are now moving to a paid subscription service and I'm not in a position to add yet another direct debit to my regular expenses! So I was proposing to move to HA to monitor and control my PW3. Based on what you have said above, and the fact that the Raspberry Pi 3b+ (rpi) has both ethernet and Wifi, I was wondering if I could use Wifi for direct connect to PW3 and ethernet for connection to my home network. Basic tests on a bare rpi, running your "pypowerwall tedapi" command over a PuTTY session (over the ethernet) seem to suggest this IS possible. If u have a moment, would appreciate your thoughts? Regards, Harry.

**Reply by jasonacox** [reply]
>  Raspberry Pi 3b+ (rpi) has both ethernet and Wifi, I was wondering if I could use Wifi for direct connect to PW3 and ethernet for connection to my home network

Yes. This is what I do and here are sone detailed instructions:  https://github.com/jasonacox/Powerwall-Dashboard/discussions/607

**Reply by Highland-Horatio** [reply]
Great stuff Jason - much appreciated. Got Grafana up and running with little or no bother - doing this as proof of concept - now moving on try and install HA. Regards, Harry.
